src/main/scala/firrtl/Compiler.scala

```diff
@@ -33,71 +33,75 @@ import java.io.Writer
 import Utils._
 import firrtl.passes._
 
-trait Compiler extends LazyLogging {
-  def run(c: Circuit, w: Writer)
+
+// ===========================================
+//                 Annotations
+// -------------------------------------------
+case class AnnotationException(message: String) extends Exception(message)
+trait Named { def name: String }
+case class ModuleName(name: String) extends Named
+case class ComponentName(name: String, module: ModuleName) extends Named
+
```

Comment:
> Using references to the actual Named IR node would actually have a smaller memory footprint

Not a good idea. All of the passes are operating under the assumption that the Firrtl graph is a tree. If there are cycles, we can introduce a whole number of errors (as well as having to deal with the HUGE headache of stale references). The cost of allocating a String is _totally_ acceptable.